vowpalwabbit/offset_tree_cont.cc

```diff
@@ -117,131 +131,144 @@ uint32_t offset_tree::predict(LEARNER::single_learner& base, example& ec)
     }
   }
   ec.l.cb = saved_label;
-  return (cur_node.id - binary_tree.internal_node_count() + 1);  // 1 to k
+  return (cur_node.id - _binary_tree.internal_node_count() + 1);  // 1 to k
 }
 
 bool compareByid(const node_cost& a, const node_cost& b) { return a.node_id < b.node_id; }
 
+void offset_tree::init_node_sets(v_array<cb_class>& ac)
+{
+  for (uint32_t i = 0; i < ac.size(); i++)
+  {
+    // Sanity check. actions are 1 based index
+    assert(ac[i].action > 0);
+    // TODO: Handle negative cost
+    assert(ac[i].cost > 0.f);  
+
+    uint32_t node_id = ac[i].action + _binary_tree.internal_node_count() - 1;
+    VW_DBG(_dd) << "otree_c: learn() ac[" << i << "].action  = " << ac[i].action << ", node_id  = " << node_id
+                << std::endl;
+    if (_binary_tree.nodes[node_id].depth < _binary_tree.depth())
+      _nodes_depth_1.push_back({node_id, ac[i].cost / ac[i].probability});
+    else
+      _nodes_depth.push_back({node_id, ac[i].cost / ac[i].probability});
+  }
+
+  if (_nodes_depth.empty())
+  {
+    _nodes_depth = _nodes_depth_1;
+    _nodes_depth_1.clear();
+  }
+
+  std::sort(_nodes_depth_1.begin(), _nodes_depth_1.end(), compareByid);
+  std::sort(_nodes_depth.begin(), _nodes_depth.end(), compareByid);
+}
+
+#define RIGHT 1
+#define LEFT -1
+
+void offset_tree::reduce_depth() {
+  // Completed processing all nodes at current depth
+  // Moving on to depth current-1
+  std::sort(_nodes_depth_1.begin(), _nodes_depth_1.end(), compareByid);
```

Comment:
We do not need to sort at each level. We just need to sort in the first iteration of the while loop, which can also be omitted if we do the following: 
In "init_node_sets" we sort _nodes_depth by increasing order (compareByid) and define a new compareByid_1 that sorts _nodes_depth_1 by decreasing order. In that case we do not need sort in reduce_depth even for the first iteration.